README.md
# sift_anatomy_extra


CHANGELOG
sift_anatomy_20140420_extra.tar.gz
 - first to contain describe_keys.c

sift_anatomy_20140530.tar.gz
 - Review pascal Getreuer part 1.
 - const variable
 - lib_util.c (max tab)..
 - assert


sift_anatomy_20140609.tar.gz
 - Review Pascal Getreuer part 2.

// 20140609
- Bug in the normalization of the feature vector.
This is not bin*255/max(bins) but bin*512/norml2(bins)

// 20140611
- sqrt(2) replace by a define
- bug in assert tests comparing floats that can be equal
  correction use FLOAT_EPSILON to include the equality case.
  in lib_descriptor in assert M_PI
- M_PI was defined twice with various number of decimals
- fmod to express the gradient in 0 2pi (atan2 gives in -pi pi)
   the problem fmod(x+10*M_PI,2*M_PI) -> fmod(x+2*M_PI,2*M_PI)
- Extrema interpolation. Possible difference between anatomy and lowe binaries.
  If the offset in scale is larger than 0.5 (the half distance between two
  scale samples) Lowe doesn't (I believe) jump to the next scale. Only, change
  sin position are implemented.  Observations: With this sampling parameters
  the threshold over the offset has an impact on the keypoint location (number
  of keypoints position) Visually, our results are more similar to Lowe's if we
  don't

// 20140616
 - in a few assert. replaced FLT_FLOAT with a larger constant EPSILON.
 - in lib_util linear_conversion : response in [0,1]. skip the normalization if the vector
   is constant (0 division)
 - array_max
 - array_min
 - find_array_max
 - fixed index in description  (s+1) -> s. This is the index used to chose the
   layer of the octace (gradient).

// 20140618
 - bug, find two first minima of an array
 - new functions in lib_util.c (euclidean_distance  euclidean_distance_square)
 - TODO, the comparison should be performed on the squared euclidean distance
   (for speed and precision).  tested it gives the same result than euclidean
   distance and exactly the same result as Lowe's code (siftDemoV4/match.c).

// 20140630
 - in lib_sift.c replace hardcoded default parameters with a call to the
   sift_assign_default_parameters function
 - in demo_extract_patch.c  remove printImage definition
 - in demo_extract_patch.c  function thumbnail_ori rewritten
 - in demo_extract_patch.c  remove not used functions gradient_field_ori() and
   gradient_field_descr()
 - sprintf() is replaced with snprintf() The functions snprintf() and
   vsnprintf() write at most size bytes (including the terminating null byte
   ('\0')) to str.
 - remove 'count' output for functions accumulating histograms,
   sift_extract_feature_vector and sift_accumulate_orientation_histogram,
   they now return void

// 20140801 - Major change in the interface
 sift_cli
 - new parameter parser pick_option() (inspired by enric's implementation).
 - each parameter is adjustable individually
 - the help menu and README are changed accordingly
 TODO:
 - add an option to do only detection
 - remove a few useless output files (extra_????.txt), namely ..REJ.. ..MULT..
 - euclidean_distance() uses euclidean_distance()
 - replace ceil by floor+1 !!! in lib_description.

// 20140808 _Leo_F_
 - new function fprintf_one_keypoint(). This will guaranty consistency between sift_cli and match_cli.
 - refactoring match_cli, the useless 'pair' structure is replaced by three sift_keypoints structure and a list of doubles
 - sift_keypoints structure: the list of pointers to keypoint structures is replaced by a list of keypoints.

// 20140819
 - fixed BUG segmentation fault (realloc_sift_keypoints - bad comparison size and capacity)
 - fixed memory leaks in filtering question (the copy of the keypoint was allocated regardeless of the test results)
 - fixed BUG affecting sift_find_ori_and_fill_descriptors and sift_anatomy_orientation_and_description:
         sx instead of sy in keypoints_attribute_one_orientation

// 20140901
 - two example codes that use the lib_sift.o: example.c and sift_cli_standard.c
   (requested by the reviewer)
 - unnecessary casts from (void*), as in the return value of malloc, can be
   avoided.
 - UNDONE printing the keypoints starting with keypoints with the largest
   scale.
 - FIXED BUG in the orientation histogram peak interpolation. The wrong formula
   led to a quantified set of principal orientations.

// 20140909
 - computation of number of octave (now put aside in a static function)
 - threshold adaptation to scalespace discretization (now put aside in a static function)


// 20140911
 - update io_png.c (it includes the string.h library that defines strcmp).
 - anatomy2lowe script to convert orientations and descriptors
 - FIXED BUG in match_cli, a segfault was raised in the absence of any match.
 - Check if all input filenames are valid


// 20140929
 - TODO in lib_keypoint.h definition of the 'val' field/ val is not normalized, remove wrong comment.

// 20141002
 - Add 'iter' in struct keypoint (number of iterations).
 - When the discrete extrema is on the edge of the image, the process is stopped.
 - TEMP, no jump in scale
 - DEBUG, nmber of iteration was zero sometimes (impoossible) 
 - The definition of the DoG operator is function of the sampling parameter ss_nspo.
   A new parameter (dog_nspo) allows to define  the DoG operator without changing the discretization (gradual)
 - In gradual, all scales in the DoG scalespace are computed from the scalespace seed ( sigma_min).

// 20141007
 - Three floating point precision (single double quad, 32 64 128) defined in
   lib_util.h preprocessor definition QUAD DOUBLE

src/gradual.cpp
/*
IPOL SIFT
Copyright (C) 2014, Ives Rey-Otero, CMLA ENS Cachan 
<ives.rey-otero@cmla.ens-cachan.fr>

Version 20140911 (September 11th, 2014)

This C ANSI source code is related to the IPOL publication

    [1] "Anatomy of the SIFT Method." 
        I. Rey Otero  and  M. Delbracio
        Image Processing Online, 2013.
        http://www.ipol.im/pub/algo/rd_anatomy_sift/

An IPOL demo is available at
        http://www.ipol.im/pub/demo/rd_anatomy_sift/





== Patent Warning and License =================================================

The SIFT method is patented 

    [3] "Method and apparatus for identifying scale invariant features
      in an image."
        David G. Lowe
        Patent number: 6711293
        Filing date: Mar 6, 2000
        Issue date: Mar 23, 2004
        Application number: 09/519,89
  
 These source codes are made available for the exclusive aim of serving as
 scientific tool to verify the soundness and completeness of the algorithm
 description. Compilation, execution and redistribution of this file may
 violate patents rights in certain countries. The situation being different
 for every country and changing over time, it is your responsibility to
 determine which patent rights restrictions apply to you before you compile,
 use, modify, or redistribute this file. A patent lawyer is qualified to make
 this determination. If and only if they don't conflict with any patent terms,
 you can benefit from the following license terms attached to this file.


This program is free software: you can use, modify and/or
redistribute it under the terms of the simplified BSD
License. You should have received a copy of this license along
this program. If not, see
<http://www.opensource.org/licenses/bsd-license.html>.

*/
/**
 * @file sift.c
 * @brief [[MAIN]] The SIFT method 
 *
 * @li basic SIFT transform applied to one image
 * @li verbose SIFT transform 
 *
 * @author Ives Rey-Otero <ives.rey-otero@cmla.ens-cachan.fr>
 */


#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <string.h>


extern "C" {

#include "lib_sift_anatomy.h"
#include "lib_io_scalespace.h"
#include "io_png.h"
#include "lib_util.h"
#include "iio.h"
#include "lib_dense_anatomy.h"

}
#include "io_exr.h"




void print_usage()
{
    fprintf(stderr, "Anatomy of the SIFT method (www.ipol.im/pub/pre/82/)  ver 20140801         \n");
    fprintf(stderr, "Usage:  sift_cli image [options...]                                        \n");
    fprintf(stderr, "                                                                           \n");
    fprintf(stderr, "   -ss_noct        (8)  number of octaves                                  \n");
    fprintf(stderr, "   -ss_nspo        (3)  number of scales per octaves                       \n");
    fprintf(stderr, "   -ss_dmin      (0.5)  the sampling distance in the first octave          \n");
    fprintf(stderr, "   -ss_smin      (0.8)  blur level on the seed image                       \n");
    fprintf(stderr, "   -ss_sin       (0.5)  assumed level of blur in the input image           \n");
    fprintf(stderr, "                                                                           \n");
    fprintf(stderr, "   -thresh_dog (0.0133) threshold over the DoG response                    \n");
    fprintf(stderr, "   -thresh_edge   (10)  threshold over the ratio of principal curvature    \n");
    fprintf(stderr, "         ########### NO DESCRIPTION  ###################                   \n");
    fprintf(stderr, "   -ori_nbins    (36)   number of bins in the orientation histogram        \n");
    fprintf(stderr, "   -ori_thresh  (0.8)   threhsold for considering local maxima in          \n");
    fprintf(stderr, "                        the orientation histogram                          \n");
    fprintf(stderr, "   -ori_lambda  (1.5)   sets how local is the analysis of the gradient     \n");
    fprintf(stderr, "                        distribution                                       \n");
    fprintf(stderr, "                                                                           \n");
    fprintf(stderr, "   -descr_nhist   (4)   number of histograms per dimension                 \n");
    fprintf(stderr, "   -descr_nori    (8)   number of bins in each histogram                   \n");
    fprintf(stderr, "   -descr_lambda  (6)   sets how local the descriptor is                   \n");
    fprintf(stderr, "                                                                           \n");
    fprintf(stderr, "   -verb_keys   label   flag to output the intermediary sets of keypoints  \n");
    fprintf(stderr, "   -verb_ss     label   flag to output the scalespaces (Gaussian and DoG)  \n");
    fprintf(stderr, "                                                                           \n");
    fprintf(stderr, "          -----  EXTRA   dense_anatomy -----                               \n");
    fprintf(stderr, "                                                                           \n");
    fprintf(stderr, "   -ss_fnspo (3.00) float number of scales per octaves (-ss_nspo not used) \n");
    fprintf(stderr, "   -flag_semigroup   BOOL (1)    semigroup (1) or direct (0)               \n");
    fprintf(stderr, "   -flag_dct         BOOL (0)    dct (1) or discrete (0)                   \n");
    fprintf(stderr, "   -flag_log         BOOL (0)    normalized Laplacian (1) or DoG (0)       \n");
    fprintf(stderr, "   -flag_interp     (0)  bilin (0) / DCT (1)/ bsplines (3,5,..,11)         \n");
    fprintf(stderr, "   -itermax             5        max number of iterations                  \n");
    fprintf(stderr, "          NOTE: if itermax=0 then all the discrete extrema are output      \n");
    fprintf(stderr, "                              no threshold is applied                      \n");
    fprintf(stderr, "   -epsilon        FLT_EPSILON    (for _myfloat comparison)                \n");
    fprintf(stderr, "   -dog_nspo    3  number of scales per octaves for DoG operator definition\n");
    fprintf(stderr, "   -ofstMax_X   (0.5)   interpolation validity domain definition in space  \n");
    fprintf(stderr, "   -ofstMax_S   (0.5)                 ... in scale                         \n");
    fprintf(stderr, "   -flag_jumpinscale   (0 in gradual / not an option in gradual)           \n");
    fprintf(stderr, "   -flag_bordereffect    BOOL (0 = standard lowe)                           \n");
}


/**
 *
 * Output 
 *   -1 : malformed argument
 *    0 : option not found  
 *    1 : option found
 */
static int pick_option(int* c, char*** v, char* opt, char* val)
{
    int output = 0;
    int argc = *c;
    char **argv = *v;
    // scan the command line for '-opt'
    for(int i = 0; i < argc; i++){
        if (argv[i][0] == '-' && 0 == strcmp(argv[i]+1, opt))
        {
            // check for a corresponding value
            if (i == argc-1){
                output = -1;
            }
            else{
                if (argv[i+1][0] == '-'){
                    output  = -1;
                }
                // if the option call is well formed
                else{
                    // copy the option value ...
                    strcpy(val, argv[i+1]);
                    // ... and remove from the command line
                    for (int j = i; j < argc - 2; j++){
                        (*v)[j] = (*v)[j+2];
                    }
                    *c -= 2;
                    output = 1;
                }
            }
            // print an error if not succes
            if(output == -1){
                fprintf(stderr, "Fatal error: option %s requires an argument.\n", opt);
            }
        }
    }
    return output;
}



static int parse_options(int argc, char** argv,
                         struct sift_parameters* p,
                         int *flag_keys,
                         int *flag_ss,
                         char* label_keys,
                         char* label_ss,
                         int *flag_semigroup,
                         int *flag_dct,
                         int *flag_log,
                         int *flag_interp,
                         int *flag_bordereffect)
{
    int isfound;
    char val[128];

    isfound = pick_option(&argc, &argv, "ss_noct", val);
    if (isfound ==  1)    p->n_oct = atoi(val);
    if (isfound == -1)    return EXIT_FAILURE;

    isfound = pick_option(&argc, &argv, "ss_nspo", val);
    if (isfound ==  1){
        p->n_spo = atoi(val);
        p->fnspo = atof(val);
    }
    if (isfound == -1)    return EXIT_FAILURE;

    isfound = pick_option(&argc, &argv, "ss_dmin", val);
    if (isfound ==  1)    p->delta_min = atof(val);
    if (isfound == -1)    return EXIT_FAILURE;

    isfound = pick_option(&argc, &argv, "ss_smin", val);
    if (isfound ==  1)    p->sigma_min = atof(val);
    if (isfound == -1)    return EXIT_FAILURE;

    isfound = pick_option(&argc, &argv, "ss_sin", val);
    if (isfound ==  1)    p->sigma_in = atof(val);
    if (isfound == -1)    return EXIT_FAILURE;

    isfound = pick_option(&argc, &argv, "thresh_dog", val);
    if (isfound ==  1)    p->C_DoG = atof(val);
    if (isfound == -1)    return EXIT_FAILURE;

    isfound = pick_option(&argc, &argv, "thresh_edge", val);
    if (isfound ==  1)    p->C_edge = atof(val);
    if (isfound == -1)    return EXIT_FAILURE;

    isfound = pick_option(&argc, &argv, "ori_nbins", val);
    if (isfound ==  1)    p->n_bins = atoi(val);
    if (isfound == -1)    return EXIT_FAILURE;

    isfound = pick_option(&argc, &argv, "ori_thresh", val);
    if (isfound ==  1)    p->t = atof(val);
    if (isfound == -1)    return EXIT_FAILURE;

    isfound = pick_option(&argc, &argv, "ori_lambda", val);
    if (isfound ==  1)    p->lambda_ori = atof(val);
    if (isfound == -1)    return EXIT_FAILURE;

    isfound = pick_option(&argc, &argv, "descr_nhist", val);
    if (isfound ==  1)    p->n_hist = atoi(val);
    if (isfound == -1)    return EXIT_FAILURE;

    isfound = pick_option(&argc, &argv, "descr_nori", val);
    if (isfound ==  1)    p->n_ori = atoi(val);
    if (isfound == -1)    return EXIT_FAILURE;

    isfound = pick_option(&argc, &argv, "descr_lambda", val);
    if (isfound ==  1)    p->lambda_descr = atof(val);
    if (isfound == -1)    return EXIT_FAILURE;

    isfound = pick_option(&argc, &argv, "verb_keys", val);
    if (isfound ==  1){
        *flag_keys = 1;
        strcpy(label_keys, val);
    }
    if (isfound == -1)    return EXIT_FAILURE;

    isfound = pick_option(&argc, &argv, "verb_ss", val);
    if (isfound ==  1){
        *flag_ss = 1;
        strcpy(label_ss, val);
    }
    if (isfound == -1)    return EXIT_FAILURE;

    // EXTRA DENSE ANATOMY
    isfound = pick_option(&argc, &argv, "flag_semigroup", val);
    if (isfound ==  1)    *flag_semigroup = atoi(val);
    if (isfound == -1)    return EXIT_FAILURE;
    isfound = pick_option(&argc, &argv, "flag_dct", val);
    if (isfound ==  1)    *flag_dct = atoi(val);
    if (isfound == -1)    return EXIT_FAILURE;
    isfound = pick_option(&argc, &argv, "flag_log", val);
    if (isfound ==  1)    *flag_log = atoi(val);
    if (isfound == -1)    return EXIT_FAILURE;
    isfound = pick_option(&argc, &argv, "itermax", val);
    if (isfound ==  1)    p->itermax = atoi(val);
    if (isfound == -1)    return EXIT_FAILURE;
    isfound = pick_option(&argc, &argv, "dog_nspo", val);
    if (isfound ==  1)    p->dog_nspo = atoi(val);
    if (isfound == -1)    return EXIT_FAILURE;
    isfound = pick_option(&argc, &argv, "epsilon", val);
    if (isfound ==  1)    p->epsilon = atof(val);
    if (isfound == -1)    return EXIT_FAILURE;
    isfound = pick_option(&argc, &argv, "flag_interp", val);
    if (isfound ==  1)    *flag_interp = atoi(val);
    if (isfound == -1)    return EXIT_FAILURE;
    isfound = pick_option(&argc, &argv, "flag_bordereffect", val);
    if (isfound ==  1)    *flag_bordereffect = atoi(val);
    if (isfound == -1)    return EXIT_FAILURE;

    isfound = pick_option(&argc, &argv, "ss_fnspo", val);
    if (isfound ==  1)    p->fnspo = atof(val);
    if (isfound == -1)    return EXIT_FAILURE;

    //  controlling the interpolation
    isfound = pick_option(&argc, &argv, "flag_jumpinscale", val);
    if (isfound ==  1)    p->flag_jumpinscale = atoi(val);
    if (isfound == -1)    return EXIT_FAILURE;
    isfound = pick_option(&argc, &argv, "ofstMax_X", val);
    if (isfound ==  1)    p->ofstMax_X = atof(val);
    if (isfound == -1)    return EXIT_FAILURE;
    isfound = pick_option(&argc, &argv, "ofstMax_S", val);
    if (isfound ==  1)    p->ofstMax_S = atof(val);
    if (isfound == -1)    return EXIT_FAILURE;

    // check for unknown option call
    for(int i = 0; i < argc; i++){
        if (argv[i][0] == '-'){
            fprintf(stderr, "Fatal error: option \"-%s\" is unknown.\n", argv[i]+1);
            print_usage();
            return EXIT_FAILURE;
        }
    }
    // check for input image
    if (argc != 2){
        print_usage();
        return EXIT_FAILURE;
    }
    return EXIT_SUCCESS;
}




// NOTE: copied from iio.c 
static bool string_suffix(const char *s, const char *suf)
{
	int len_s = strlen(s);
	int len_suf = strlen(suf);
	if (len_s < len_suf)
		return false;
	return 0 == strcmp(suf, s + (len_s - len_suf));
}

// reads the filename suffix and uses the appropriate routine in function.
static _myfloat* read_image(char* n, int* w, int* h)
{
    float* x;

    if (string_suffix(n, ".PNG") || string_suffix(n, ".png")){
        // IO_PNG
        size_t wt, ht;
        x = io_png_read_f32_gray(n, &wt, &ht);
        for(unsigned int i=0; i < wt*ht; i++)
            x[i] /= 256.0;
        *w = wt; *h = ht;
    }
    else if ( string_suffix(n, ".EXR") || string_suffix(n, ".exr")){
        // IO_EXR
        x = readEXR_float(n, w, h);
    }
    else{
        fatal_error("This version only recognizes png and exr(32bits) images");
    }

    _myfloat* y = (_myfloat*)malloc((*w)*(*h)*sizeof(_myfloat));
    for (int i = 0; i < (*w)*(*h); i++){
        y[i] = (_myfloat)x[i];
    }
    free(x);
    return(y);
}



// Print function
// Note that the value is normalized (equivalent to dog_nspo = 3)
// Note that the value in the 27 long vector are not normalized
void print_keypoints_and_vals(const struct sift_keypoints* keys, int dog_nspo)
{
   // float k_nspo =  exp( M_LN2/( (float)(dog_nspo)));
   // float k_3 =  exp( M_LN2/( (float)3));
    float k_nspo = pow(2, (float)(dog_nspo));
    float k_3 =  pow(2, 3.0);
    float factor = (k_3 - 1) / (k_nspo - 1);

    for (int i = 0; i < keys->size; i++){
        struct keypoint* k = keys->list[i];
        float val = k->val * factor;
#ifdef QUAD // conversion to string and then fprintf
        char str[1024]; // long strinpl
        quadmath_snprintf(str, sizeof(str), "%.30Qg %.30Qg %.30Qg %.30Qg %.30Qg",  k->x, k->y, k->sigma, val);
        fprintf(stdout, "%s \n", str);
#else
        fprintf(stdout, "%f %f %f %f ", k->x, k->y, k->sigma, val);
#endif
        //
        // The 27 points in the cube.
        for(int n = 0; n < 27; n++){
            fprintf(stdout, "%33.30f ", k->neighbors[n]);
        }
        // EXTRA - the interpolation offset for the last interpolation and grid position
        // k->s is always equal to 1 for the 'gradual' implementation // TODO correct the structure somewhere before output.
        fprintf(stdout, "%i %i %i %f %f %f ", k->i, k->j, k->s,  k->ofstX,  k->ofstY,  k->ofstS);
        fprintf(stdout, "\n");
    }
}







/** @brief Main SIFT routine
 * 
 * takes one image as input.        
 * outputs the extracted keypoints in the standard output.
 * 
 * @param flag for the SIFT transform
 * 
 * 0 = one image -> one txt file 
 * 1 = one image -> all txt files
 * 2 = one image -> all txt files + scalespace and DoG
 * 
 */
int main(int argc, char **argv)
{
    // Setting default parameters
    struct sift_parameters* p = sift_assign_default_parameters();
    int flagverb_keys = 0;
    int flagverb_ss = 0;
    char label_ss[256];
    char label_keys[256];
    strcpy(label_ss, "extra");
    strcpy(label_keys, "extra");
    // EXTRA DENSE
    int flag_semigroup = 1;
    int flag_dct = 1;
    int flag_log = 0;
    int flag_interp = 0;
    int flag_bordereffect = 0;

    // Parsing command line
    //int res = parse_options(argc, argv, p, &flagverb_keys, &flagverb_ss, label_keys, label_ss);
  //  int res = parse_options(argc, argv, p, &flagverb_keys, &flagverb_ss, label_keys, label_ss,
  //                                    &flag_semigroup, &flag_dct, &flag_log);
    int res = parse_options(argc, argv, p, &flagverb_keys, &flagverb_ss, label_keys, label_ss,
                                      &flag_semigroup, &flag_dct, &flag_log, &flag_interp, &flag_bordereffect);
    if (res == EXIT_FAILURE)
        return EXIT_FAILURE;

    /** Loading image */
    int w, h;
    _myfloat* x = read_image(argv[1], &w, &h);
    if(!x)
        fatal_error("File \"%s\" not found.", argv[1]);


    /** Memory dynamic allocation */
    // WARNING 6 steps of the algorithm are recorded.
    struct sift_keypoints **kk = (sift_keypoints **)xmalloc(6*sizeof(struct sift_keypoints*));
    for(int i = 0; i < 6; i++)
        kk[i] = sift_malloc_keypoints();
    // WARNING 4 scale-space representation are recorded (Gaussian, Laplacian, two gradient components)
    struct sift_scalespace **ss = (sift_scalespace **)xmalloc(4*sizeof(struct sift_scalespace*));

    /** Algorithm */
 //   struct sift_keypoints* k = sift_anatomy_gradual(x, w, h, p, ss, kk, flag_semigroup, flag_dct, flag_log);
    struct sift_keypoints* k;
    if (flag_bordereffect == 0){
        k = sift_anatomy_gradual(x, w, h, p, ss, kk, flag_semigroup, flag_dct, flag_log, flag_interp);
    }
    else{
        k = sift_anatomy_gradual_auxil(x, w, h, p, ss, kk, flag_semigroup, flag_dct, flag_log, flag_interp);
    }

    /** OUTPUT */
    //int flag = flagverb_keys + 1;
    //sift_print_keypoints(k, flag);
    //sift_print_keypoints(k, 0);
    //print_keypoints_and_vals(k, p->n_spo);
    print_keypoints_and_vals(k, p->dog_nspo);
    // With normalized value (consistent with DoG threshold for nspo = 3

   // char name[FILENAME_MAX];
    if(flagverb_keys == 1){
       // sprintf(name,"extra_NES_%s.txt",label_keys);              sift_save_keypoints(kk[0], name, 0);
       // sprintf(name,"extra_DoGSoftThresh_%s.txt",label_keys);    sift_save_keypoints(kk[1], name, 0);
       // sprintf(name,"extra_ExtrInterp_%s.txt",label_keys);       sift_save_keypoints(kk[2], name, 0);
       // sprintf(name,"extra_DoGThresh_%s.txt",label_keys);        sift_save_keypoints(kk[3], name, 0);
       // sprintf(name,"extra_OnEdgeResp_%s.txt",label_keys);       sift_save_keypoints(kk[4], name, 0);
       // sprintf(name,"extra_FarFromBorder_%s.txt",label_keys);    sift_save_keypoints(kk[5], name, 0);
    }
    if (flagverb_ss == 1){
       // sprintf(name,"scalespace_%s",label_ss);     print_sift_scalespace_gray_nearestneighbor(ss[0],name);
       // sprintf(name,"DoG_%s",label_ss);            print_sift_scalespace_rgb(ss[1],name);
    }

    /* memory deallocation */
    xfree(x);
    xfree(p);
    sift_free_keypoints(k);
    for(int i = 0; i < 6; i++){
        sift_free_keypoints(kk[i]);   // TODO pour gradual
    }
    xfree(kk);
    for(int i = 0; i < 4; i++){
       // sift_free_scalespace(ss[i]);   // TODO pour gradual
    }
    xfree(ss);


    return EXIT_SUCCESS;
}

